src/crypto/index.js

```diff
@@ -236,10 +281,368 @@ Crypto.prototype.init = async function() {
+ * you can define which methods you want to use, "&" with the appopriate mask,
+ * then test for truthiness.  Or if you want to just trust everything, then use
+ * the value alone.  However, I wonder if bit masks are too obscure...
+ */
+Crypto.prototype.checkUserTrust = function(userId) {
+    const userCrossSigning = this._deviceList.getStoredCrossSigningForUser(userId);
+    if (!userCrossSigning) {
+        return 0;
+    }
+    return this._crossSigningInfo.checkUserTrust(userCrossSigning) << 1;
+};
+
+/**
+ * Check whether a given device is trusted.
+ *
+ * @param {string} userId The ID of the user whose devices is to be checked.
+ * @param {string} deviceId The ID of the device to check
+ *
+ * @returns {integer} a bit mask indicating how the user is trusted (if at all)
+ *  - returnValue & 1: device marked as verified
+ *  - returnValue & 2: trust-on-first-use cross-signing key
+ *  - returnValue & 4: user's cross-signing key is verified and device is signed
+ *
+ * TODO: see checkUserTrust
+ */
+Crypto.prototype.checkDeviceTrust = function(userId, deviceId) {
+    let rv = 0;
+
+    const device = this._deviceList.getStoredDevice(userId, deviceId);
+    if (device && device.isVerified()) {
+        rv |= 1;
+    }
+    const userCrossSigning = this._deviceList.getStoredCrossSigningForUser(userId);
+    if (device && userCrossSigning) {
+        rv |= this._crossSigningInfo.checkDeviceTrust(userCrossSigning, device) << 1;
+    }
+    return rv;
+};
+
+/*
+ * Event handler for DeviceList's userNewDevices event
+ */
+Crypto.prototype._onDeviceListUserCrossSigningUpdated = async function(userId) {
+    if (userId === this._userId) {
+        await this._checkOwnCrossSigningTrust();
+    } else {
+        await this._checkDeviceVerifications(userId);
+    }
+};
+
+/*
+ * Check the copy of our cross-signing key that we have in the device list and
+ * see if we can get the private key. If so, mark it as trusted.
+ */
+Crypto.prototype._checkOwnCrossSigningTrust = async function() {
+    const userId = this._userId;
+
+    // If we see an update to our own master key, check it against the master
+    // key we have and, if it matches, mark it as verified
+
+    // First, get the new cross-signing info
+    const newCrossSigning = this._deviceList.getStoredCrossSigningForUser(userId);
+    if (!newCrossSigning) {
+        logger.error(
+            "Got cross-signing update event for user " + userId +
+            " but no new cross-signing information found!",
+        );
+        return;
+    }
+
+    const seenPubkey = newCrossSigning.getId();
+    const changed = this._crossSigningInfo.getId() !== seenPubkey;
+    let privkey;
+    if (changed) {
+        // try to get the private key if the master key changed
+        logger.info("Got new master key", seenPubkey);
+
+        let error;
+        do {
+            privkey = await new Promise((resolve, reject) => {
+                this._baseApis.emit("cross-signing.newKey", {
```

Comment:
maybe the way to go with this is, once the getkey callback is a thing, do that, then emit newKey saying whether it's trusted or not